### DISCUSSION TITLE
Using eth_signTypedData_v4 with Ethereum Provider returns undefined

### DISCUSSION THREAD

**Discussion by wavywtf** [WalletConnect/walletconnect-monorepo#3290] [Developer Support] [answered]
I am using a vanilla js/html setup with `Ethereum Provider` and I am running into an issue where `eth_signTypedData_v4` returns undefined. If I switch my code to `window.ethereum` (I'm on desktop) I am able to use the same `eth_signTypedData_v4` function without any issues. 

My Ethereum Provider code:
```
var ethereumProvider = null;
var web3 = null;
async function init() { 
try {	  
		if (!ethereumProvider) {		 
			ethereumProvider = await pkg.EthereumProvider.init({       
				projectId,       
				showQrModal: true,      
				qrModalOptions: { 
					themeMode: "light",  
				chains: [5],				
				methods: ["eth_sendTransaction", "eth_sendRawTransaction", "personal_sign", "eth_signTypedData_v4"],       
				events: ['chainChanged', 'accountsChanged', 'connect', 'disconnect'],
			});

			if(!ethereumProvider.session) {
			await ethereumProvider.connect();
			}

			web3 = new Web3(ethereumProvider);

	} catch (err) {   
		console.error(err); 
	}
}

```
And the eth_signTypedData_v4 part of my code:
```
	web3.currentProvider.sendAsync({        
			method: "eth_signTypedData_v4",            
			params: [account, data],            
			from: account            
		},function(err, result) {        
			if (err) {            
				$("#response").text(JSON.parse(JSON.stringify(err.message)));             
				return;                
			}               
			const signature = result.result.substring(2);  
});
```

Which returns:
`Cannot read properties of undefined (reading 'substring')`

Any ideas on how to fix this?

**Comment by glitch-txs** [comment]
Hi wavy, which wallet are you testing it with?

**Reply by wavywtf** [reply]
I'm on desktop, so first I tested using metamask's browser extension on desktop using window.ethereum "default" code, and that worked as intended. Then using the ethereum provider code, I used the qr modal to connect with my wallet on the metamask app on my phone and that returned the undefined error.

**Reply by glitch-txs** [reply]
HI wavy I wasn't able to reproduce, we're using this app for testing https://react-app.walletconnect.com/
Do you have any public repo we can take a look at?

**Reply by wavywtf** [reply]
no, I'm justing testing with the code that I put in my question, using your [eth provider repo code from the previous discussion](https://github.com/glitch-txs/ethereum-provider-cdn). Also the linked site in your reply doesn't seem to be related to my current setup.

**Reply by glitch-txs** [reply]
seems to be related to web3js, keep in mind init is aync so you might need a getter like `let getWeb3 = ()=>web3` and call that inside other functions to get the current provider, not sure if this is the case though.

**Reply by wavywtf** [reply]
using the same setup/code with window.ethereum for getting/setting a provider, instead of ethereum provider init(), works fine with it-
```
if (window.ethereum) {	
		let web3 = new Web3(window.ethereum);	  
		window.web3 = web3;		
		window.provider = await detectEthereumProvider()
		provider = await detectEthereumProvider()
}
```

**Reply by wavywtf** [reply]
@glitch-txs I created a [gist of the code](https://gist.github.com/wavywtf/b65f0b0298f9c5c13eace44a42a77c61) in basic form, where I am getting the result as undefined. Hopefully that helps clarify things.

**Reply by glitch-txs** [reply, marked as answer]
Thanks for this, seems like with desktop it returns an object and with mobile return the signature itself

This is how I'd handle it
```ts
let signature;		
if(result.result){
  signature = result.result.substring(2);
}else{
  signature = result.substring(2);
}
console.log(signature)
```
